be/src/storage/horizontal_compaction_task.cpp

```diff
@@ -0,0 +1,150 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+#include "storage/horizontal_compaction_task.h"
+
+#include <vector>
+
+#include "column/schema.h"
+#include "common/statusor.h"
+#include "runtime/current_thread.h"
+#include "storage/olap_common.h"
+#include "storage/rowset/rowset.h"
+#include "storage/rowset/rowset_writer.h"
+#include "storage/storage_engine.h"
+#include "storage/tablet.h"
+#include "storage/vectorized/chunk_helper.h"
+#include "storage/vectorized/tablet_reader.h"
+#include "storage/vectorized/tablet_reader_params.h"
+#include "util/time.h"
+#include "util/trace.h"
+
+namespace starrocks {
+
+Status HorizontalCompactionTask::run_impl() {
+    Statistics statistics;
+    RETURN_IF_ERROR(_horizontal_compact_data(&statistics));
+
+    TRACE_COUNTER_INCREMENT("merged_rows", statistics.merged_rows);
+    TRACE_COUNTER_INCREMENT("filtered_rows", statistics.filtered_rows);
+    TRACE_COUNTER_INCREMENT("output_rows", statistics.output_rows);
+
+    RETURN_IF_ERROR(_validate_compaction(statistics));
+    TRACE("[Compaction] horizontal compaction validated");
+
+    _commit_compaction();
+    TRACE("[Compaction] horizontal compaction committed");
+
+    return Status::OK();
+}
+
+Status HorizontalCompactionTask::_horizontal_compact_data(Statistics* statistics) {
+    TRACE("[Compaction] start horizontal comapction data");
+    // 1: init
+    int64_t max_rows_per_segment = CompactionUtils::get_segment_max_rows(
+            config::max_segment_file_size, _task_info.input_rows_num, _task_info.input_rowsets_size);
+
+    std::unique_ptr<RowsetWriter> output_rs_writer;
+    RETURN_IF_ERROR(CompactionUtils::construct_output_rowset_writer(
+            _tablet.get(), max_rows_per_segment, _task_info.algorithm, _task_info.output_version, &output_rs_writer));
+
+    vectorized::Schema schema = vectorized::ChunkHelper::convert_schema_to_format_v2(_tablet->tablet_schema());
+    vectorized::TabletReader reader(std::static_pointer_cast<Tablet>(_tablet->shared_from_this()),
+                                    output_rs_writer->version(), schema);
+    vectorized::TabletReaderParams reader_params;
+    reader_params.reader_type = ReaderType::READER_LEVEL_COMPACTION;
+    reader_params.profile = _runtime_profile.create_child("merge_rowsets");
+
+    int32_t chunk_size = CompactionUtils::get_read_chunk_size(
+            config::compaction_memory_limit_per_worker, config::vector_chunk_size, _task_info.input_rows_num,
+            _task_info.input_rowsets_size, _task_info.segment_iterator_num);
+    VLOG(1) << "compaction task_id:" << _task_info.task_id << ", tablet=" << _tablet->tablet_id()
```

Comment:
What does the task_id is used for? What message do you want to give the user?

Reply:
the task_id is unique for each compaction task. It has the following benefits:
1. useful to locate compaction task in log for debug
2. can now how many compaction tasks there is
3. can be used to cancel the specific compaction task in the future.